lib/service/chat_service.dart

```diff
@@ -0,0 +1,223 @@
+                        .map((e) => app.Message.fromJson(e))
+                        .toList();
+                for (var element in _listeners) {
+                  element.onResponseMessageReturnPayload(
+                      newMessages, response.id);
+                }
+              } catch (e) {
+                log.info("[CHAT page] RESP error: $e");
+              }
+            }
+            break;
+          default:
+            break;
+        }
+      },
+      onDone: () async {
+        log.info("[CHAT] onDone");
+
+        Future.delayed(const Duration(seconds: 5), () async {
+          if (_address != null && _id != null && _wallet != null) {
+            _address = null;
+            _id = null;
+            _wallet = null;
+            log.info("[CHAT] _websocketChannel reconnecting");
+            await connect(
+              address: _address!,
+              id: _id!,
+              wallet: _wallet!,
+            );
+            for (var element in _listeners) {
+              addListener(element);
+            }
+            for (var element in _listeners) {
+              element.onDoneCalled.call();
+            }
+          }
+        });
+      },
+    );
+  }
+
+  @override
+  void addListener(ChatListener listener) {
+    if (listener is PostcardChatListener) {
+      _listeners.add(listener);
+    }
+  }
+
+  @override
+  Future<void> dispose() async {
+    log.info("[CHAT] disconnect");
+    _address = null;
+    _id = null;
+    _wallet = null;
+    await _websocketChannel?.sink.close();
+    _websocketChannel = null;
+  }
+
+  @override
+  void sendMessage(dynamic message) {
+    _websocketChannel?.sink.add(message);
+  }
+
+  Future<Map<String, dynamic>> _getHeader(
+      String link, Pair<WalletStorage, int> wallet, String address) async {
+    final Map<String, dynamic> header = {};
+    final timestamp = DateTime.now().millisecondsSinceEpoch;
+    header["X-Api-Timestamp"] = timestamp;
+    final canonicalString = List<String>.of([
+      link,
+      "",
+      timestamp.toString(),
+    ]).join("|");
+    final hmacSha256 = Hmac(sha256, utf8.encode(Environment.chatServerHmacKey));
+    final digest = hmacSha256.convert(utf8.encode(canonicalString));
+    final sig = bytesToHex(digest.bytes);
+    header["X-Api-Signature"] = sig;
+
+    final pubKey = await wallet.first.getTezosPublicKey(index: wallet.second);
+    final authSig = await injector<TezosService>().signMessage(wallet.first,
+        wallet.second, Uint8List.fromList(utf8.encode(timestamp.toString())));
+    final token = await injector<ChatAuthService>().getAuthToken({
+      "address": address,
+      "public_key": pubKey,
+      "signature": authSig,
+      "timestamp": timestamp
+    });
+    header["Authorization"] = "Bearer $token";
+    return header;
+  }
+
+  @override
+  void removeListener() {
+    _listeners.removeLast();
+    if (_listeners.isEmpty) {
+      dispose();
+    }
+  }
+}
+
+abstract class ChatListener {}
```

Comment:
I think we don't need to create an abstract class for ChatListener. Nothing to abstract in here.